File: intellij-plugin/core/src/test/kotlin/com/decoy/core/PanelControllerTest.kt
package com.decoy.core

import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.io.File

private const val TEST_KEY = "nhtXQTnXQzKW_0m_5r-ffK8dcl03BLSzJOOdmF8bDk4="
private val ENV = mapOf("DECOY_ENCRYPTION_KEY" to TEST_KEY)

private class FakeHost(override val rootDir: File, private val confirmResult: Boolean = true) : PanelHost {
    val updates = mutableListOf<Pair<List<RequestGroup>, OverridesFile>>()
    var confirmMessages = mutableListOf<String>()

    override fun updateView(groups: List<RequestGroup>, overrides: OverridesFile) {
        updates.add(groups to overrides)
    }

    override fun confirmDestructive(message: String): Boolean {
        confirmMessages.add(message)
        return confirmResult
    }
}

private fun entry(
    id: String,
    requestId: String,
    sessionId: String,
    field: String = "EMAIL",
) = AuditEntry(id, requestId, sessionId, "2026-08-30T21:15:03.482Z", "prompt_text", field, "masked", "r", "regex")

private fun writeVaultFixture(root: File, data: Map<String, Map<String, String>>) {
    File(root, ".decoy").mkdirs()
    File(root, VAULT_RELATIVE_PATH).writeText(encryptToken(Json.encodeToString(data), TEST_KEY))
}

private fun readVaultFixture(root: File): Map<String, Map<String, String>> {
    val decrypted = decryptToken(File(root, VAULT_RELATIVE_PATH).readText(), TEST_KEY)!!
    return Json.decodeFromString(decrypted)
}

class PanelControllerTest {

    @Test
    fun `refresh pushes current audit log and overrides to the host`(@TempDir tmp: File) {
        writeAuditLog(tmp, AuditLogFile(1, listOf(entry("e1", "req-1", "session-1"))), ENV)
        withEnv(ENV) {
            val host = FakeHost(tmp)
            PanelController(host).refresh()
            assertEquals(1, host.updates.size)
            assertEquals(1, host.updates[0].first.size)
        }
    }

    @Test
    fun `clearSession clears audit and vault for that session only, no confirmation needed`(@TempDir tmp: File) {
        writeAuditLog(
            tmp,
            AuditLogFile(1, listOf(entry("e1", "r1", "session-a"), entry("e2", "r2", "session-b", "PHONE"))),
            ENV,
        )
        writeVaultFixture(
            tmp,
            mapOf(
                "session-a" to mapOf("alice@example.com" to "fake-a@example.net"),
                "session-b" to mapOf("bob@example.com" to "fake-b@example.net"),
            ),
        )
        withEnv(ENV) {
            val host = FakeHost(tmp)
            PanelController(host).clearSession("session-a")

            assertEquals(mapOf("session-b" to mapOf("bob@example.com" to "fake-b@example.net")), readVaultFixture(tmp))
            val lastGroups = host.updates.last().first
            assertEquals(1, lastGroups.size)
            assertEquals("session-b", lastGroups[0].sessionId)
            assertTrue(host.confirmMessages.isEmpty()) // per-session clear doesn't need a confirm dialog
        }
    }

    @Test
    fun `clearSessionDataOnly does nothing when the host declines confirmation`(@TempDir tmp: File) {
        writeAuditLog(tmp, AuditLogFile(1, listOf(entry("e1", "r1", "s1"))), ENV)
        withEnv(ENV) {
            val host = FakeHost(tmp, confirmResult = false)
            PanelController(host).clearSessionDataOnly()
            assertTrue(host.updates.isEmpty())
        }
    }

    @Test
    fun `clearSessionDataOnly clears audit and vault for every session but KEEPS overrides`(@TempDir tmp: File) {
        writeAuditLog(tmp, AuditLogFile(1, listOf(entry("e1", "r1", "s1"))), ENV)
        writeVaultFixture(tmp, mapOf("s1" to mapOf("a@example.com" to "b@example.net")))
        val overrides = OverridesFile(always_mask = OverrideRules(field_names = listOf("employee_id")))
        writeOverrides(tmp, overrides)

        withEnv(ENV) {
            val host = FakeHost(tmp, confirmResult = true)
            PanelController(host).clearSessionDataOnly()

            assertEquals(emptyMap<String, Map<String, String>>(), readVaultFixture(tmp))
            assertEquals(overrides, readOverrides(tmp)) // kept
            assertTrue(host.updates.last().first.isEmpty())
            assertEquals(overrides, host.updates.last().second)
        }
    }

    @Test
    fun `clearAll removes EVERYTHING when confirmed, including overrides -- no unstated exception`(@TempDir tmp: File) {
        writeAuditLog(tmp, AuditLogFile(1, listOf(entry("e1", "r1", "s1"))), ENV)
        writeVaultFixture(tmp, mapOf("s1" to mapOf("a@example.com" to "b@example.net")))
        writeOverrides(
            tmp,
            OverridesFile(
                always_mask = OverrideRules(field_names = listOf("employee_id")),
                never_mask = OverrideRules(field_names = listOf("status")),
            ),
        )

        withEnv(ENV) {
            val host = FakeHost(tmp, confirmResult = true)
            PanelController(host).clearAll()

            assertEquals(emptyMap<String, Map<String, String>>(), readVaultFixture(tmp))
            assertEquals(emptyOverridesFile(), readOverrides(tmp))
            assertTrue(host.updates.last().first.isEmpty())
            assertEquals(emptyOverridesFile(), host.updates.last().second)
        }
    }

    @Test
    fun `saveOverrides persists and is reflected in the next update`(@TempDir tmp: File) {
        withEnv(ENV) {
            val host = FakeHost(tmp)
            val overrides = OverridesFile(always_mask = OverrideRules(field_names = listOf("employee_id")))
            PanelController(host).saveOverrides(overrides)

            assertEquals(overrides, readOverrides(tmp))
            assertEquals(overrides, host.updates.last().second)
        }
    }
}

/**
 * JVM processes can't mutate real environment variables per-test safely,
 * so key loading in these tests goes through the `env` parameter directly
 * wherever the code under test accepts one. `withEnv` exists only for
 * readability at call sites that don't thread it explicitly (readAuditLog
 * etc. inside PanelController always use System.getenv() by default) --
 * this project passes DECOY_ENCRYPTION_KEY via the JVM's real environment
 * when running these tests (see core/build.gradle.kts test task) so
 * PanelController's calls into readAuditLog/writeAuditLog/etc, which have
 * no way to receive a custom `env` from PanelHost, still resolve the key.
 */
private inline fun withEnv(env: Map<String, String>, block: () -> Unit) {
    require(System.getenv("DECOY_ENCRYPTION_KEY") == env["DECOY_ENCRYPTION_KEY"]) {
        "Test JVM must be launched with DECOY_ENCRYPTION_KEY=$TEST_KEY " +
            "(configured in core/build.gradle.kts's test task) for PanelControllerTest to pass."
    }
    block()
}

File: vscode-extension/src/panelController.test.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";

import { writeAuditLog } from "./auditLog";
// eslint-disable-next-line @typescript-eslint/no-require-imports
import fernet = require("fernet");
import { PanelController, PanelHost } from "./panelController";
import { readOverrides, writeOverrides } from "./overrides";

const ENV = { DECOY_ENCRYPTION_KEY: "nhtXQTnXQzKW_0m_5r-ffK8dcl03BLSzJOOdmF8bDk4=" } as NodeJS.ProcessEnv;

function makeTmpDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "decoy-panel-test-"));
}

function makeFakeHost(rootDir: string, confirmResult = true): PanelHost & { htmlHistory: string[] } {
  const htmlHistory: string[] = [];
  return {
    rootDir,
    cspSource: "vscode-webview://test",
    codiconsUri: "vscode-webview://test/codicon.css",
    generateNonce: () => "fixed-nonce",
    postHtml: (html: string) => {
      htmlHistory.push(html);
    },
    confirmDestructive: async () => confirmResult,
    htmlHistory,
  };
}

function writeVaultFixture(root: string, data: Record<string, Record<string, string>>): void {
  fs.mkdirSync(path.join(root, ".decoy"), { recursive: true });
  const secret = new fernet.Secret(ENV.DECOY_ENCRYPTION_KEY as string);
  const token = new fernet.Token({ secret }).encode(JSON.stringify(data));
  fs.writeFileSync(path.join(root, ".decoy", "vault.enc"), token, "utf8");
}

function readVaultFixture(root: string): Record<string, Record<string, string>> {
  const raw = fs.readFileSync(path.join(root, ".decoy", "vault.enc"), "utf8");
  const secret = new fernet.Secret(ENV.DECOY_ENCRYPTION_KEY as string);
  const token = new fernet.Token({ secret, token: raw, ttl: 0 });
  return JSON.parse(token.decode() as string);
}

describe("PanelController.refresh", () => {
  it("posts HTML reflecting the current audit log and overrides", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      {
        format_version: 1,
        entries: [
          {
            id: "e1",
            request_id: "req-1",
            session_id: "session-1",
            timestamp: "2026-08-30T21:15:03.482Z",
            source: "prompt_text",
            field: "EMAIL",
            decision: "masked",
            reason: "matched EMAIL regex",
            layer: "regex",
          },
        ],
      },
      ENV,
    );
    const host = makeFakeHost(root);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      new PanelController(host).refresh();
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    expect(host.htmlHistory).toHaveLength(1);
    expect(host.htmlHistory[0]).toContain("EMAIL");
  });
});

describe("PanelController.handleMessage", () => {
  it("clearSession clears both the audit log entry and the vault mapping for that session", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      {
        format_version: 1,
        entries: [
          {
            id: "e1",
            request_id: "req-1",
            session_id: "session-a",
            timestamp: "2026-08-30T21:15:03.482Z",
            source: "prompt_text",
            field: "EMAIL",
            decision: "masked",
            reason: "matched EMAIL regex",
            layer: "regex",
          },
          {
            id: "e2",
            request_id: "req-2",
            session_id: "session-b",
            timestamp: "2026-08-30T21:16:00.000Z",
            source: "prompt_text",
            field: "PHONE",
            decision: "masked",
            reason: "matched PHONE regex",
            layer: "regex",
          },
        ],
      },
      ENV,
    );
    writeVaultFixture(root, {
      "session-a": { "alice@example.com": "fake-a@example.net" },
      "session-b": { "bob@example.com": "fake-b@example.net" },
    });

    const host = makeFakeHost(root);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      await new PanelController(host).handleMessage({ command: "clearSession", sessionId: "session-a" });
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    expect(readVaultFixture(root)).toEqual({ "session-b": { "bob@example.com": "fake-b@example.net" } });
    const lastHtml = host.htmlHistory[host.htmlHistory.length - 1];
    expect(lastHtml).toContain("PHONE");
    expect(lastHtml).not.toContain("session-a");
  });

  it("clearSessionDataOnly does nothing if the host declines confirmation", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      { format_version: 1, entries: [{ id: "e1", request_id: "r1", session_id: "s1", timestamp: "t", source: "prompt_text", field: "EMAIL", decision: "masked", reason: "r", layer: "regex" }] },
      ENV,
    );
    const host = makeFakeHost(root, /* confirmResult */ false);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      await new PanelController(host).handleMessage({ command: "clearSessionDataOnly" });
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    // no refresh should have been posted at all since nothing changed
    expect(host.htmlHistory).toHaveLength(0);
  });

  it("clearSessionDataOnly removes audit/vault data for every session but KEEPS overrides", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      { format_version: 1, entries: [{ id: "e1", request_id: "r1", session_id: "s1", timestamp: "t", source: "prompt_text", field: "EMAIL", decision: "masked", reason: "r", layer: "regex" }] },
      ENV,
    );
    writeVaultFixture(root, { s1: { "a@example.com": "b@example.net" } });
    const overrides = { always_mask: { patterns: [], field_names: ["employee_id"] }, never_mask: { patterns: [], field_names: [] } };
    writeOverrides(root, overrides);

    const host = makeFakeHost(root, true);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      await new PanelController(host).handleMessage({ command: "clearSessionDataOnly" });
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    expect(readVaultFixture(root)).toEqual({});
    expect(readOverrides(root)).toEqual(overrides); // kept, not cleared
    expect(host.htmlHistory[host.htmlHistory.length - 1]).toContain("No requests yet");
    expect(host.htmlHistory[host.htmlHistory.length - 1]).toContain("employee_id");
  });

  it("clearAll does nothing if the host declines confirmation", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      { format_version: 1, entries: [{ id: "e1", request_id: "r1", session_id: "s1", timestamp: "t", source: "prompt_text", field: "EMAIL", decision: "masked", reason: "r", layer: "regex" }] },
      ENV,
    );
    const host = makeFakeHost(root, /* confirmResult */ false);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      await new PanelController(host).handleMessage({ command: "clearAll" });
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    // no refresh should have been posted at all since nothing changed
    expect(host.htmlHistory).toHaveLength(0);
  });

  it("clearAll removes EVERYTHING when confirmed, including overrides -- no unstated exception", async () => {
    const root = makeTmpDir();
    writeAuditLog(
      root,
      { format_version: 1, entries: [{ id: "e1", request_id: "r1", session_id: "s1", timestamp: "t", source: "prompt_text", field: "EMAIL", decision: "masked", reason: "r", layer: "regex" }] },
      ENV,
    );
    writeVaultFixture(root, { s1: { "a@example.com": "b@example.net" } });
    writeOverrides(root, {
      always_mask: { patterns: [], field_names: ["employee_id"] },
      never_mask: { patterns: [], field_names: ["status"] },
    });

    const host = makeFakeHost(root, true);
    process.env.DECOY_ENCRYPTION_KEY = ENV.DECOY_ENCRYPTION_KEY;
    try {
      await new PanelController(host).handleMessage({ command: "clearAll" });
    } finally {
      delete process.env.DECOY_ENCRYPTION_KEY;
    }

    expect(readVaultFixture(root)).toEqual({});
    expect(readOverrides(root)).toEqual({
      always_mask: { patterns: [], field_names: [] },
      never_mask: { patterns: [], field_names: [] },
    });
    const lastHtml = host.htmlHistory[host.htmlHistory.length - 1];
    expect(lastHtml).toContain("No requests yet");
    expect(lastHtml).not.toContain("employee_id");
    expect(lastHtml).not.toContain("status");
  });

  it("saveOverrides persists and is reflected in the next refresh", async () => {
    const root = makeTmpDir();
    const host = makeFakeHost(root);
    const overrides = {
      always_mask: { patterns: [], field_names: ["employee_id"] },
      never_mask: { patterns: [], field_names: [] },
    };
    await new PanelController(host).handleMessage({ command: "saveOverrides", overrides });

    expect(readOverrides(root)).toEqual(overrides);
    expect(host.htmlHistory[host.htmlHistory.length - 1]).toContain("employee_id");
  });
});
